exp/operator/controllers/genericprovider_controller.go

```diff
@@ -92,14 +104,175 @@ func (r *GenericProviderReconciler) Reconcile(ctx context.Context, req reconcile
 		return result, err
 	}
 
-	return ctrl.Result{}, nil
+	reader, err := r.secretReader(ctx, genericProvider)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	cfg, err := configclient.New("", configclient.InjectReader(reader))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	providerConfig, err := cfg.Providers().Get(genericProvider.GetName(), util.ClusterctlProviderType(genericProvider))
+	if err != nil {
+		conditions.Set(genericProvider, conditions.FalseCondition(
+			operatorv1.PreflightCheckCondition,
+			operatorv1.UnknownProviderReason,
+			v1alpha4.ConditionSeverityWarning,
+			fmt.Sprintf(unknownProviderMessage, genericProvider.GetName()),
+		))
+		return ctrl.Result{}, nil
+	}
+
+	clusterClient := cluster.New(cluster.Kubeconfig{}, cfg)
+
+	// ensure the custom resource definitions required by clusterctl are in place
+	if err := clusterClient.ProviderInventory().EnsureCustomResourceDefinitions(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	spec := genericProvider.GetSpec()
+
+	var repositoryClient repository.Client
+	if spec.FetchConfig != nil && spec.FetchConfig.Selector != nil {
+		var cmRepo repository.Repository
+		cmRepo, err = r.configmapRepository(ctx, genericProvider)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		repositoryClient, err = repository.New(providerConfig, cfg, repository.InjectRepository(cmRepo))
+	} else {
+		repositoryClient, err = repository.New(providerConfig, cfg)
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// TODO how is the user to define the target namespace?
+	// If we use the below method, we will never be able to have default namespaces.
+	options := repository.ComponentsOptions{
+		TargetNamespace:   genericProvider.GetNamespace(),
+		WatchingNamespace: "",
+		SkipVariables:     false,
+	}
+	if spec.Version != nil {
+		options.Version = *spec.Version
+	}
+
+	components, err := repositoryClient.Components().Get(options)
+	if err != nil {
+		conditions.Set(genericProvider, conditions.FalseCondition(
+			v1alpha4.ReadyCondition,
+			operatorv1.ComponentsGatherErrorReason,
+			v1alpha4.ConditionSeverityWarning,
+			err.Error(),
+		))
+
+		return ctrl.Result{}, err
+	}
+
+	objs, err := customizeDeployment(genericProvider, components.InstanceObjs())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	components.SetInstanceObjs(objs)
+	installer := clusterClient.ProviderInstaller()
+	installer.Add(components)
+
+	if certManagerRequired(components) {
+		// Before installing the providers, ensure the cert-manager Webhook is in place.
+		certManager, err := clusterClient.CertManager()
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// NOTE: this can block for a while..
+		if err := certManager.EnsureInstalled(); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if util.IsInfrastructureProvider(genericProvider) && genericProvider.GetName() == "metal3" {
+		// HACK alert: for some reason metal3 webhooks are installed in capi-webhook-system,
```

Comment:
is there some way to avoid provider-specific hacks?

Reply:
of course, i'll debug an try and see what is happening